kinesis_consumer.cpp

```diff
@@ -0,0 +1,266 @@
+/*-------------------------------------------------------------------------
+ *
+ * kinesis_consumer.cpp
+ *	  Implementation for kinesis consumer C wrapper
+ *
+ * Copyright (c) 2013-2016, PipelineDB
+ *
+ *-------------------------------------------------------------------------
+ */
+#include <aws/kinesis/KinesisClient.h>
+#include <aws/core/client/ClientConfiguration.h>
+#include <aws/core/auth/AWSCredentialsProviderChain.h>
+#include <aws/kinesis/model/GetRecordsResult.h>
+#include <aws/kinesis/model/GetShardIteratorResult.h>
+#include <aws/kinesis/model/GetShardIteratorRequest.h>
+#include <aws/core/utils/Outcome.h>
+#include <aws/core/utils/logging/AWSLogging.h>
+#include <aws/core/utils/logging/DefaultLogSystem.h>
+#include <aws/kinesis/model/GetRecordsRequest.h>
+#include <aws/core/client/RetryStrategy.h>
+#include <aws/core/utils/logging/FormattedLogSystem.h>
+#include <aws/core/client/DefaultRetryStrategy.h>
+
+#include "kinesis_consumer.h"
+#include "conc_queue.hpp"
+
+#include <unordered_set>
+#include <functional>
+#include <thread>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/time.h>
+
+using namespace Aws::Auth;
+using namespace Aws::Client;
+using namespace Aws::Kinesis;
+using namespace Aws::Kinesis::Model;
+using namespace Aws::Utils;
+
+struct kinesis_consumer
+{
+	KinesisClient *client;
+	concurrent_queue<GetRecordsOutcome*> *queue;
+	std::atomic<bool> keep_running;
+	std::thread *thread;
+	Aws::String shard_iter;
+};
+
+// Override the default logging system to print to stderr
+class AltLogSystem : public Logging::FormattedLogSystem
+{
+  public:
+	AltLogSystem(Logging::LogLevel logLevel) : FormattedLogSystem(logLevel) {}
+	virtual ~AltLogSystem() {}
+
+  protected:
+	virtual void ProcessFormattedStatement(Aws::String&& statement)
+	{
+		fprintf(stderr, "%s", statement.c_str());
+	}
+};
+
+static double
+get_time()
+{
+    struct timeval tv;
+    gettimeofday(&tv, 0);
+    return tv.tv_sec + (tv.tv_usec / 1000000.0);
+}
+
+static void consume_thread(kinesis_consumer *kc);
+
+// create the state for the consumer.
+// note - this is a blocking call wrt obtaining the shard iterator.
+// on error this returns NULL.
+kinesis_consumer*
+kinesis_consumer_create()
+{
+	Aws::Utils::Logging::InitializeAWSLogging(Aws::MakeShared<AltLogSystem>("logging", Aws::Utils::Logging::LogLevel::Error));
+
+	ClientConfiguration config;
+	config.region = Aws::Region::US_WEST_2;
+	config.verifySSL = false;
+	
+	config.retryStrategy = Aws::MakeShared<DefaultRetryStrategy>("kinesis_consumer", 5, 25);
+	auto *kc = new KinesisClient(config);
+
+	GetShardIteratorRequest request;
+
+	request.SetStreamName("test");
+	request.SetShardId("shardId-000000000000");
```

Comment:
This needs to know how to automatically consume from all shards, as in `pipeline_kafka`:

https://github.com/pipelinedb/pipeline_kafka/blob/master/pipeline_kafka.c#L714